icu4c/source/i18n/unitconverter.cpp

```diff
@@ -5,14 +5,521 @@
+
+        factorNum = std::pow(factorNum, std::abs(power));
+        factorDen = std::pow(factorDen, std::abs(power));
+
+        if (shouldFlip) {
+            // Flip Numerator and Denomirator.
+            std::swap(factorNum, factorDen);
+        }
+    }
+
+    // Flip the `Factor`, for example, factor= 2/3, flippedFactor = 3/2
+    void flip() {
+        std::swap(factorNum, factorDen);
+
+        for (int i = 0; i < CONSTANTS_COUNT; i++) {
+            constants[i] *= -1;
+        }
+    }
+
+    // Apply SI prefix to the `Factor`
+    void applySiPrefix(UMeasureSIPrefix siPrefix) {
+        if (siPrefix == UMeasureSIPrefix::UMEASURE_SI_PREFIX_ONE) return; // No need to do anything
+
+        double siApplied = std::pow(10.0, std::abs(siPrefix));
+
+        if (siPrefix < 0) {
+            factorDen *= siApplied;
+            return;
+        }
+
+        factorNum *= siApplied;
+    }
+};
+
+//////////////////////////
+/// BEGIN DATA LOADING ///
+//////////////////////////
+
+class UnitConversionRatesSink : public ResourceSink {
+  public:
+    explicit UnitConversionRatesSink(Factor *conversionFactor) : conversionFactor(conversionFactor) {}
+
+    void put(const char *key, ResourceValue &value, UBool /*noFallback*/,
+             UErrorCode &status) U_OVERRIDE {
+        ResourceTable conversionRateTable = value.getTable(status);
+        if (U_FAILURE(status)) { return; }
+
+        for (int32_t i = 0; conversionRateTable.getKeyAndValue(i, key, value); ++i) {
+            StringPiece keySP(key);
+
+            if (keySP == "factor") {
+                value.getUnicodeString(status);
+
+            }
+
+            else if (keySP == "offset") {
+                value.getUnicodeString(status);
+            }
+
+            else if (keySP == "target") {
+                // TODO(younies): find a way to convert UnicodeStirng to StringPiece
+                // conversionRate->target.set(value.getUnicodeString(status));
+            }
+
+            if (U_FAILURE(status)) { return; }
+        }
+    }
+
+  private:
+    Factor *conversionFactor;
+};
+
+ConvertUnit extractConvertUnit(StringPiece source, UErrorCode &status) {
+    ConvertUnit result;
+    // TODO(hugovdm): implement.
+    //   NOTE: use the database.
+    //         UnitConversionRatesSink sink(&conversionFactor);
+    //         ures_getAllItemsWithFallback(rb.getAlias(), key.data(), sink, status);
+    //
+    //   NOTE:
+    //     LocalUResourceBundlePointer rb(ures_openDirect(nullptr, "units", &status));
+    //     CharString key;
+    //     key.append("convertUnit/", status);
+    //     key.append(source, status);
+    return result;
+}
+
+/*/
+ * Add single factor element to the `Factor`. e.g "ft3m", "2.333" or "cup2m3". But not "cup2m3^3".
+ */
+void addSingleFactorConstant(Factor &factor, StringPiece baseStr, int32_t power, Signal signal) {
+    if (baseStr == "ft_to_m") {
+        factor.constants[CONSTANT_FT2M] += power * signal;
```

Comment:
Ping

Reply:
Shane, how data-driven do we want these eventually? Constant values are in the resources - we could load/parse them at run-time. If we do so, do we also want the number of constants to be loaded at runtime? Meaning the resource bundle could change to add new constants, and the conversion rates could refer to those constants, without anything needing to be recompiled.

On the otherhand, we do have some generated code, how reasonable is the idea to deal with things at compile time, possibly via more generated code? (I wouldn't want too many diverse steps for importing CLDR data - ldml2icu is one set of steps that we connect to easily. If we do generate more code, I would want to see it be easier to do than some of the other regeneration instructions I've seen?)

Reply:
We could load the constants in a UInitOnce like I suggested we do for the full list of simple unit strings.  We could alternatively add them to the code gen Peter made for some of the other constants in measunit.cpp.  Those are both reasonable options.  If you want to add these to code gen, rather than data loading logic, read the instructions page:

http://site.icu-project.org/design/formatting/measureformat/updating-measure-unit

Reply:
You are still missing this TODO.  Please add it.

Reply:
I checked, there's a "TODO" now - who should mark this as "Resolved"?